File: July-LeetCoding-Challenge-2020/Week3/203-Remove-Linked-List-Elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 // Solution #1 Two Pointers
 // TC: O(length of nodes)
 // SC: O(1)
 // Runtime: 28 ms, faster than 60.73%
 // Memory Usage: 13.5 MB, less than 61.47%
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        
        if(!head)
            return head;
        
        while(head){
            if(head->val != val)
                break;
            head = head->next;
        }
        
        if(!head || !head->next)
            return head;
        
        ListNode *slow, *fast;
        slow = head;
        fast = head->next;
        
        while(fast){
            if(fast->val == val){
                fast = fast->next;
                slow->next = fast;
            }else{
                slow = slow->next;
                fast = fast->next;
            }
            
        }
        
        return head;
    }
};

File: July-LeetCoding-Challenge-2020/Week3/203-remove-linked-list-elements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 * Runtime: 96 ms, faster than 36.21% of JavaScript online submissions for Remove Linked List Elements.
 * Memory Usage: 41.7 MB, less than 5.15% of JavaScript online submissions for Remove Linked List Elements.
 * TC: O(N), where N is length of head
 * SC: O(1)
 */
var removeElements = function(head, val) {
    let prev, headCopy = head;

    while (head) {
        if (head.val === val) {
            if (!prev) {
                headCopy = headCopy.next;
            } else {
                prev.next = head.next;
            }
        } else {
            prev = head;
        }

        head = head.next;
    }

    return headCopy;
};